### PR TITLE
BUG: Add upstream gh-2667 for cmake paths

### LINKED ISSUE #196: Bug with imported Cmake target 1.14.0
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

The error in the automated OpenMEEG rebuild PR https://github.com/conda-forge/openmeeg-feedstock/pull/42 seems to indicate a problem with HDF5 feedstock:

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=654485&view=logs&j=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&t=841356e0-85bb-57d8-dbbc-852e683d1642&l=271

```
Running CMAKE
-- The C compiler identification is GNU 11.3.0
...
-- Found HDF5: $PREFIX/lib/libhdf5.so;$PREFIX/lib/libcrypto.so;$PREFIX/lib/libcurl.so;$BUILD_PREFIX/x86_64-conda-linux-gnu/sysroot/usr/lib/librt.so;$BUILD_PREFIX/x86_64-conda-linux-gnu/sysroot/usr/lib/libpthread.so;$PREFIX/lib/libsz.so;$PREFIX/lib/libz.so;$BUILD_PREFIX/x86_64-conda-linux-gnu/sysroot/usr/lib/libdl.so;$BUILD_PREFIX/x86_64-conda-linux-gnu/sysroot/usr/lib/libm.so (found version "1.14.0")  
...
-- Configuring done
CMake Error in OpenMEEGMaths/CMakeLists.txt:
  Imported target "HDF5::HDF5" includes non-existent path

    "/home/conda/feedstock_root/build_artifacts/hdf5_1675740981212/work/src/H5FDsubfiling"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```
Note that my [manual migration PR to 1.12.2 went smoothly](https://github.com/conda-forge/openmeeg-feedstock/pull/40) and we haven't changed OpenMEEG infra/code at all, so it seems like a new problem with 1.14.0.

### Installed packages

```shell
Copied from https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=654485&view=logs&j=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&t=841356e0-85bb-57d8-dbbc-852e683d1642&l=271


    _libgcc_mutex:            0.1-conda_forge         conda-forge
    _openmp_mutex:            4.5-2_gnu               conda-forge
    binutils_impl_linux-64:   2.39-he00db2b_1         conda-forge
    binutils_linux-64:        2.39-h5fc0e48_11        conda-forge
    bzip2:                    1.0.8-h7f98852_4        conda-forge
    c-ares:                   1.18.1-h7f98852_0       conda-forge
    ca-certificates:          2022.12.7-ha878542_0    conda-forge
    cmake:                    3.25.2-h077f3f9_0       conda-forge
    expat:                    2.5.0-h27087fc_0        conda-forge
    gcc_impl_linux-64:        11.3.0-hab1b70f_19      conda-forge
    gcc_linux-64:             11.3.0-he6f903b_11      conda-forge
    gfortran_impl_linux-64:   11.3.0-he34c6f7_19      conda-forge
    gfortran_linux-64:        11.3.0-h3c55166_11      conda-forge
    gxx_impl_linux-64:        11.3.0-hab1b70f_19      conda-forge
    gxx_linux-64:             11.3.0-hc203a17_11      conda-forge
    kernel-headers_linux-64:  2.6.32-he073ed8_15      conda-forge
    keyutils:                 1.6.1-h166bdaf_0        conda-forge
    krb5:                     1.20.1-h81ceb04_0       conda-forge
    ld_impl_linux-64:         2.39-hcc3a1bd_1         conda-forge
    libcurl:                  7.87.0-hdc1c0ab_0       conda-forge
    libedit:                  3.1.20191231-he28a2e2_2 conda-forge
    libev:                    4.33-h516909a_1         conda-forge
    libgcc-devel_linux-64:    11.3.0-h210ce93_19      conda-forge
```
```


### Environment info

```shell
Standard conda-forge build, see https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=654485&view=logs&j=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&t=841356e0-85bb-57d8-dbbc-852e683d1642&l=271
```

### PR DESCRIPTION
**Pull request by larsoner**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Used `git format-patch -1 b77cb39` (commit from https://github.com/HDFGroup/hdf5/pull/2667) then removed the release notes change from the diff.

Closes #196

### CODE REVIEW

**Comment by larsoner** [comment]
@conda-forge-admin, please rerender

**Comment by larsoner** [comment]
@hmaarrfk backported as a patch as requested. Should be green, just waiting (hours?) for the `aarch64` build to finish

File: recipe/patches/2667-Fix-improper-include-of-build-directory.patch [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+index 7c8c958bf..d5f46da75 100644
+--- a/configure.ac
++++ b/configure.ac
+@@ -3012,8 +3012,7 @@ SUBFILING_VFD=no
+ HAVE_MERCURY="no"
+
+ ## Always include subfiling directory so public header files are available
+-CPPFLAGS="$CPPFLAGS -I$ac_abs_confdir/src/H5FDsubfiling"
+-AM_CPPFLAGS="$AM_CPPFLAGS -I$ac_abs_confdir/src/H5FDsubfiling"
++H5_CPPFLAGS="$H5_CPPFLAGS -I$ac_abs_confdir/src/H5FDsubfiling"
```
Comment (hmaarrfk) [inline recipe/patches/2667-Fix-improper-include-of-build-directory.patch:22]: This is not enough. While it is OK for the git repo, their sources have the configure file pregenerated. I believe you need to patch the pregenerated configure file itself (not the configure.ac).

In either case (maybe I am wrong with my thoughts above), maybe you can inspect the outputs locally and ensure that the problematic variable is removed.

File: recipe/patches/2667-Fix-improper-include-of-build-directory.patch [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+     mercury_inc="$mercury_dir/src/util"
+
+-    CPPFLAGS="$CPPFLAGS -I$mercury_inc"
+-    AM_CPPFLAGS="$AM_CPPFLAGS -I$mercury_inc"
++    H5_CPPFLAGS="$H5_CPPFLAGS -I$mercury_inc"
+
+     HAVE_STDATOMIC_H="yes"
+     AC_CHECK_HEADERS([stdatomic.h],,[HAVE_STDATOMIC_H="no"])
+--- a/configure
++++ b/configure
```
Comment (larsoner) [inline recipe/patches/2667-Fix-improper-include-of-build-directory.patch:24]: @hmaarrfk I am a bit out of my depth, but I added the corresponding changes to `configure`. Do we think this is enough? I'm not sure an easy way to test that this actually has the desired effect...

**Comment by larsoner** [comment]
Okay in 73830b7 I disabled the new patch completely and get the [same failures](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=690102&view=logs&jobId=664cebe1-e2a9-5f60-02ec-2148f58cc324&j=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&t=841356e0-85bb-57d8-dbbc-852e683d1642) as in 1a80476, so I think 1a80476 was working. I've reverted the "patch disable" in 38d4094, so I think this is now ready again @hmaarrfk !

File: recipe/patches/2667-Fix-improper-include-of-build-directory.patch (hunk 2 above)
Comment (hmaarrfk) [reply recipe/patches/2667-Fix-improper-include-of-build-directory.patch:24]: My test for this kind of stuff is "build locally, and manually check". If it has been approved upstream it builds confidence that we don't need a longer term test

**Comment by hmaarrfk** [comment]
Hmmmm. Other ci failures?

**Comment by larsoner** [comment]
One is a download failure, the `*_mpimpi*` four were present on [73830b7](https://github.com/conda-forge/hdf5-feedstock/pull/200/commits/73830b7b966464536980dfcba361a369e0100c1a) where I disabled the patch completely, which I think means they now somehow exist on `main` as well. Could open a quick empty commit PR to check...

**Comment by larsoner** [comment]
https://github.com/conda-forge/hdf5-feedstock/pull/201

**Comment by larsoner** [comment]
Same failures -- other than the one download timeout here -- in the no-op (other than bumping version number) PR #201. So somehow those are now failing on `main` :disappointed:

**Comment by hmaarrfk** [comment]
yeah, i'm really not sure. the fact that linux is failing too is strange.

It is also strange that openssl 1.1.1 is still being built. do we really want this?

**Comment by minrk** [comment]
These issues have been fixed in 1.14.1 (#202)

**Comment by minrk** [comment]
I misread the hdf5 release notes - this was _not_ included in 1.14.1, so we should still apply the patch if we want to build with cmake

**Comment by larsoner** [comment]
That's a bummer! For what it's worth I never actually tested the version generated by this PR. So it would be good to do that if you know how. (It would be my first time so would take a while for me to figure it out.)

**Comment by minrk** [comment]
Sorry, I got myself all mixed up testing this. It is in 1.14.1 and I can verify that h5cc doesn't include the output. We can close. I'll add a cmake test in a separate PR so we can more easily verify it.

**Comment by larsoner** [comment]
Great, thanks for investigating @minrk !

**Pull request closed without merging**